executor/executor_write.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/util/types"
+)
+
+var (
+	_ Executor = &UpdateExec{}
+)
+
+// UpdateExec represents a update executor.
+type UpdateExec struct {
+	SelectExec  Executor
+	OrderedList []*ast.Assignment
+
+	updatedRowKeys map[string]bool
+	ctx            context.Context
+
+	rows        []*Row          // The rows fetched from TableExec.
+	newRowsData [][]interface{} // The new values to be set.
+	fetched     bool
+	cursor      int
+}
+
+// Next implements Execution Next interface.
+func (e *UpdateExec) Next() (*Row, error) {
+	if !e.fetched {
+		err := e.fetchRows()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		e.fetched = true
+	}
+	if e.cursor >= len(e.rows) {
+		return nil, nil
+	}
+	if e.updatedRowKeys == nil {
+		e.updatedRowKeys = map[string]bool{}
+	}
+	row := e.rows[e.cursor]
+	newData := e.newRowsData[e.cursor]
+	for _, entry := range row.RowKeys {
+		tbl := entry.Tbl
+		offset := e.getTableOffset(tbl)
+		k := entry.Key
+		oldData := row.Data[offset : offset+len(tbl.Cols())]
+		newTableData := newData[offset : offset+len(tbl.Cols())]
+
+		_, ok := e.updatedRowKeys[k]
+		if ok {
+			// Each matching row is updated once, even if it matches the conditions multiple times.
+			continue
+		}
+
+		// Update row
+		handle, err1 := tables.DecodeRecordKeyHandle(kv.Key(k))
+		if err1 != nil {
+			return nil, errors.Trace(err1)
+		}
+
+		err1 = e.updateRecord(handle, oldData, newTableData, tbl, offset, false)
+		if err1 != nil {
+			return nil, errors.Trace(err1)
+		}
+		e.updatedRowKeys[k] = true
+	}
+	e.cursor++
+	return &Row{}, nil
+}
+
+func (e *UpdateExec) fetchRows() error {
+	for {
+		row, err := e.SelectExec.Next()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if row == nil {
+			return nil
+		}
+		data := make([]interface{}, len(e.SelectExec.Fields()))
+		newData := make([]interface{}, len(e.SelectExec.Fields()))
+		for i, f := range e.SelectExec.Fields() {
+			data[i] = f.Expr.GetValue()
+			newData[i] = data[i]
+			if e.OrderedList[i] != nil {
+				val, err := evaluator.Eval(e.ctx, e.OrderedList[i].Expr)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				newData[i] = val
+			}
+		}
+		row.Data = data
+		e.rows = append(e.rows, row)
+		e.newRowsData = append(e.newRowsData, newData)
+	}
+}
+
+func (e *UpdateExec) getTableOffset(t table.Table) int {
+	fields := e.SelectExec.Fields()
+	i := 0
+	for i < len(fields) {
+		field := fields[i]
+		if field.Table.Name.L == t.TableName().L {
+			return i
+		}
+		i += len(field.Table.Columns)
+	}
+	return 0
+}
+
+func (e *UpdateExec) updateRecord(h int64, oldData, newData []interface{}, t table.Table, offset int, onDuplicateUpdate bool) error {
+	if err := t.LockRow(e.ctx, h); err != nil {
+		return errors.Trace(err)
+	}
+
+	cols := t.Cols()
+	touched := make(map[int]bool, len(cols))
+
+	assignExists := false
+	var newHandle interface{}
+	for i, asgn := range e.OrderedList {
+		if asgn == nil {
```

Comment:
the OrderedList has the same length as result fields, but only contains assign list in the update statement..